File: 1D_DP/LC_322.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.12.06

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1, INT_MAX);
        dp[0] = 0;
        for (int val = 1; val <= amount; val++){
            for(int coin: coins){
                if (val - coin >= 0 && dp[val-coin] != INT_MAX){
                    dp[val] = min(dp[val], dp[val-coin] + 1);
                }
            }
        }
        if (dp[amount] == INT_MAX){
            return -1;
        }
        return dp[amount];

    }
};

File: 1D_DP/LC_322.py
# Written by: Jian-Yuan Lin
# Date: 2022.12.06

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float('inf')] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount+1):
            for c in coins:
                if i >= c:
                    dp[i] = min(dp[i], dp[i-c] + 1)
        
        return dp[amount] if dp[amount] != float('inf') else -1